services/ui/src/lib/components/VariableDiag/Histogram.svelte

```diff
@@ -0,0 +1,87 @@
+<script>
+  import { format, scaleLinear, min, max, extent } from "d3";
+
+  export let width = 0;
+  export let height = 0;
+
+  export let bins = [];
+
+  let margin = {
+    top: 0,
+    right: 0,
+    bottom: 32,
+    left: 0,
+  };
+
+  let formatNum = format("~s");
+
+  $: domain = [min(bins, (d) => d.lower), max(bins, (d) => d.upper)];
+  $: range = extent(bins, (d) => d.value);
+
+  $: x = scaleLinear(domain, [margin.left, width - margin.left - margin.right]);
+  $: y = scaleLinear(range, [height - margin.top - margin.bottom, margin.bottom]);
+
+  $: bars = bins
+    .map((bin) => ({
+      x: x(bin.lower),
+      y: y(bin.value),
+      width: x(bin.upper) - x(bin.lower),
+      height: y(0) - y(bin.value),
+    }))
+    .filter((bar) => bar.height >= 1);
```

Comment:
Checking my understanding on d3 - this eliminates bars from the diagram that are set to a height of 1, correct?

Reply:
Close. This is actually vanilla JavaScript (`map` and `filter` are both methods on JavaScript’s `Array`). This use of `filter` keeps any bar whose `height` is at least 1, and discards any bars whose `height` is less than 1, because there are a lot of bins, and we don’t need to create a `<rect>` for anything that is going to be too small to see—that just bogs down the DOM.

`d3` is involved in calculating the height and width (and the x- and y-coordinates of the lower left corner of the rectangle). I have two linear scales that transform a value from our diagnostic data into SVG coordinates. So `x(0)` returns the x-coordinate for the value `0` in our SVG.